bad requests to etcdserver cause panic

If I curl the raft endpoint with an empty body, etcd happily panics:

```
panic: invalid transition [leader -> candidate]

goroutine 20 [running]:
runtime.panic(0x668720, 0xc208000b30)
    /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/raft.(*raft).becomeCandidate(0xc208034210)
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:291 +0x70
github.com/coreos/etcd/raft.(*raft).campaign(0xc208034210)
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:321 +0x35
github.com/coreos/etcd/raft.(*raft).Step(0xc208034210, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:339 +0xa4
github.com/coreos/etcd/raft.(*Node).run(0xc208046900, 0xc208034210)
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:122 +0x366
created by github.com/coreos/etcd/raft.Restart
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:74 +0x21b

goroutine 16 [IO wait]:
net.runtime_pollWait(0x7ff1e1f1dd48, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20802a300, 0x72, 0x0, 0x0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20802a300, 0x0, 0x0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc20802a2a0, 0x7dfd30, 0x0, 0x7ff1e1f1c440, 0xb)
    /usr/local/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*TCPListener).AcceptTCP(0xc208032048, 0x45aa43, 0x0, 0x0)
    /usr/local/go/src/pkg/net/tcpsock_posix.go:234 +0x5d
net/http.tcpKeepAliveListener.Accept(0xc208032048, 0x0, 0x0, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:1947 +0x4b
net/http.(*Server).Serve(0xc208004720, 0x7ff1e1f1ddf8, 0xc208032048, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:1698 +0x91
net/http.(*Server).ListenAndServe(0xc208004720, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:1688 +0x14d
net/http.ListenAndServe(0x73de50, 0x5, 0x0, 0x0, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:1778 +0x79
main.main()
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/main.go:52 +0xb2

goroutine 19 [finalizer wait]:
runtime.park(0x414670, 0x904890, 0x902b29)
    /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x904890, 0x902b29)
    /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
    /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
    /usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 22 [select]:
github.com/coreos/etcd/etcdserver.(*Server).run(0xc20802a230)
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:59 +0x3dd
created by github.com/coreos/etcd/etcdserver.Start
    /home/jon/git/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:54 +0x9d

goroutine 23 [IO wait]:
net.runtime_pollWait(0x7ff1e1f1dc98, 0x72, 0x0)
    /usr/local/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20802a680, 0x72, 0x0, 0x0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20802a680, 0x0, 0x0)
    /usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc20802a620, 0xc20805f000, 0x1000, 0x1000, 0x0, 0x7ff1e1f1c440, 0xb)
    /usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc208032018, 0xc20805f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
    /usr/local/go/src/pkg/net/net.go:122 +0xe7
net/http.(*liveSwitchReader).Read(0xc20804e328, 0xc20805f000, 0x1000, 0x1000, 0xf8e, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:206 +0xaf
io.(*LimitedReader).Read(0xc20803c0e0, 0xc20805f000, 0x1000, 0x1000, 0x1d, 0x0, 0x0)
    /usr/local/go/src/pkg/io/io.go:399 +0xd0
bufio.(*Reader).fill(0xc208004240)
    /usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
bufio.(*Reader).ReadSlice(0xc208004240, 0x10806200a, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/pkg/bufio/bufio.go:298 +0x22c
bufio.(*Reader).ReadLine(0xc208004240, 0x0, 0x0, 0x0, 0x422d00, 0x0, 0x0)
    /usr/local/go/src/pkg/bufio/bufio.go:326 +0x69
net/textproto.(*Reader).readLineSlice(0xc208022d50, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/pkg/net/textproto/reader.go:55 +0x9d
net/textproto.(*Reader).ReadLine(0xc208022d50, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/pkg/net/textproto/reader.go:36 +0x4e
net/http.ReadRequest(0xc208004240, 0xc208028b60, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/request.go:556 +0xc7
net/http.(*conn).readRequest(0xc20804e300, 0x0, 0x0, 0x0)
    /home/jon/go/src/pkg/net/http/server.go:577 +0x276
net/http.(*conn).serve(0xc20804e300)
    /home/jon/go/src/pkg/net/http/server.go:1132 +0x61e
created by net/http.(*Server).Serve
    /home/jon/go/src/pkg/net/http/server.go:1721 +0x313
```

This is because the empty request body gets unmarshalled into an empty `raftpb.Message`. The default value of `raftpb.Message.Type` is 0.. which corresponds to `msgHup`.. which causes `raft.Node.Step` to campaign.. which causes it to `becomeCandidate`... and then if it's the leader, it will panic, since this is an invalid state transition.

Clearly this should be fixed.

--- a/raft/node.go
+++ b/raft/node.go
@@ -205,16 +205,25 @@ func (n *node) Tick() {
 }
 
 func (n *node) Campaign(ctx context.Context) error {
-	return n.Step(ctx, pb.Message{Type: msgHup})
+	return n.step(ctx, pb.Message{Type: msgHup})
 }
 
 func (n *node) Propose(ctx context.Context, data []byte) error {
-	return n.Step(ctx, pb.Message{Type: msgProp, Entries: []pb.Entry{{Data: data}}})
+	return n.step(ctx, pb.Message{Type: msgProp, Entries: []pb.Entry{{Data: data}}})
+}
+
+func (n *node) Step(ctx context.Context, m pb.Message) error {
+	// ignore unexpected local messages receiving over network
+	if m.Type == msgHup || m.Type == msgBeat {
+		// TODO: return an error?
+		return nil
+	}
+	return n.step(ctx, m)
 }
 
 // Step advances the state machine using msgs. The ctx.Err() will be returned,
 // if any.
-func (n *node) Step(ctx context.Context, m pb.Message) error {
+func (n *node) step(ctx context.Context, m pb.Message) error {
 	ch := n.recvc
 	if m.Type == msgProp {
 		ch = n.propc

--- a/raft/node_test.go
+++ b/raft/node_test.go
@@ -18,7 +18,8 @@ func TestNodeStep(t *testing.T) {
 			propc: make(chan raftpb.Message, 1),
 			recvc: make(chan raftpb.Message, 1),
 		}
-		n.Step(context.TODO(), raftpb.Message{Type: int64(i)})
+		msgt := int64(i)
+		n.Step(context.TODO(), raftpb.Message{Type: msgt})
 		// Proposal goes to proc chan. Others go to recvc chan.
 		if int64(i) == msgProp {
 			select {
@@ -27,10 +28,18 @@ func TestNodeStep(t *testing.T) {
 				t.Errorf("%d: cannot receive %s on propc chan", i, mtmap[i])
 			}
 		} else {
-			select {
-			case <-n.recvc:
-			default:
-				t.Errorf("%d: cannot receive %s on recvc chan", i, mtmap[i])
+			if msgt == msgBeat || msgt == msgHup {
+				select {
+				case <-n.recvc:
+					t.Errorf("%d: step should ignore msgHub/msgBeat", i, mtmap[i])
+				default:
+				}
+			} else {
+				select {
+				case <-n.recvc:
+				default:
+					t.Errorf("%d: cannot receive %s on recvc chan", i, mtmap[i])
+				}
 			}
 		}
 	}

